### DISCUSSION TITLE
Is it an abuse of C#10 ?

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/csharplang#5317] [Q&A]
After C#10 introduced `global using`, I found it quite interesting for fixing some generic type parameters which should stay the same for the whole project.

```CSharp
public static class Foo<TGlobal1>
{
    public static class Bar<TGlobal2>
    {
         public class SomeUsefulClass<TOther1, TOther2>
         {
             TGlobal1 Get<TOther1, TOther2>(TGlobal2 global2, TOther1 val1, TOther2 val2)
             {
                 return default;
             }
         }
    }
}
```

and then you do

```
global using static Foo1<string>.Foo2<int>;
```

I think, this might be useful for library authors. However, the names of the classes are somewhat ugly.

**Comment by huoyaoyuan** [comment]
What's the benefits comparing to simpler syntax then?

**Reply by voroninp** [reply]
Simple example.

Let's imagine one designs `Result<TSuccess, TError>` to use functional style for error propagation.  And you do not want to constrain `TError` type, but it's likely that the shape of the error will be the same for the whole project.
If you define Result in a classic way, then clients of that code will be forced to use the pair of types, or inherit from it `Result<TScuccess> : Result<TSuccess, string> {}`. 

However, it's not always desired:
```CSharp
class A<TError> {}
class A<TSuccess, TError> : A<TError>{}
class B : A<string>{}
class B<TSuccess>: A<TSuccess, string>{}
``` 
Here `B<TSuccess>` is not `B`. One must introduce implicit conversion from former to latter. And sometimes the hierarchy of classes should be fixed/sealed.

You could argue that one can use interfaces for this, but then you have problems with implicit conversions. That's why it is `ActionResult<T>` instead of `IActionResult<T>`

So the alternative would be:

```CSharp
public static class Results
{
    public static class HaveErrorOfType<TError>
    {
        public class Result {...}
        
        public class Result<TSuccess> : Result {...}

        // And here you may define utility methods which will be immediately available when 'global using static Results.HaveErrorOfType<string>;' is used.
    }
}
```

**Reply by huoyaoyuan** [reply]
It's hard to say abusing in customer code. If there's no undefined behavior, doing anything should be OK.
